contribute/design/1.x/cstor/pool/pool_expansion_by_expanding_disk.md

```diff
@@ -0,0 +1,134 @@
+
+- BlockDevice Replacement In Pool:
+    cstor-pool manager will iterate over all the BlockDevices present in CstorPoolInstance and verifies whether the current blockdevice is for BlockDevice replacement request[How cstor-pool-manager can know? We are maintaining annotation called openebs.io/predecessor: <replaced_blockdevice_name> on the claim of that blockdevice which will be populated by admission server on replacement request]? If requested BlockDevice is valid for replacement then the pool manager will perform BlockDevice replacement operation on the pool.
+
+### Proposed Implementations
+
+- CStor-pool Manager inbuilt CSPI controller watches for CSPI resources and tries to get the desired state via every sync operation. As part of reconciliation logic, it will iterate over all the BlockDevices that exist on CStorPoolInstance resource if it found BlockDevice resource capacity(considerable changes i.e difference is > 500MB) is greater than the existing capacity of BlockDevice on CstorPoolInstanceBlockDevice then we will follow below-mentioned steps to perform the [pool expansion](#Pool_Expansion_Steps).
+- [What if cStor is consuming part of partitioned disk/entire partitioned disk? The User has to extend the partition manually after that NDM will update the capacity on the corresponding blockdevice CR. Once NDM updates the capacity on BlockDevice in next reconciliation of CSPI Pool Expansion will be triggered].
+- [What if only one blockdevice got expanded in case of raid/mirror configuration? Pool expansion operation will be triggered but there won't be any change in the size of the pool].
+
+NOTE: During the cStor pool creation time and pool import time `autoexpand` property will be configured to `on`(configuring during import time will support for pools which were migrated from SPC to CSPC).
+
+Currently CstorPoolInstanceBlockDevice holds below configurations 
+```go
+// CStorPoolInstanceBlockDevice contains the details of block devices that
+// constitutes a raid group.
+type CStorPoolInstanceBlockDevice struct {
+	// BlockDeviceName is the name of the block device.
+	BlockDeviceName string `json:"blockDeviceName"`
+
+	// Capacity is the capacity of the block device.
+    // This field will be populated by CSPI controller.
+	Capacity uint64 `json:"capacity"`
+
+	// DevLink is the dev link for block devices.
+    // This field will be populated by CSPI controller.
+	DevLink string `json:"devLink"`
+}
+```
+
+### Steps to perform user stories
+
+- User has to make sure that blockdevice capacity needs to be updated.
+
+#### Pool Expansion Steps:
+
+Step1: Execute `zpool online -e <pool_name> <disk_name>` [Why zpool online command? This will help to fix `GPT PMBR size mismatch` complaints raised by kernel].
+
+Step2: Reload the partition table by executing `partprobe <path_to_disk>`
+
+Step3: Inform the pool saying disk was expanded by triggering following command
```

Comment:
step 1 and 3 is same. why we are firing zpool online command 2 times?

Reply:
Triggering first-time helps me to get rid of below error
```sh
GPT PMBR size mismatch (1310719 != 1835007) will be corrected by w(rite).
Disk /dev/sdb: 7 GiB, 7516192768 bytes, 1835008 sectors
Units: sectors of 1 * 4096 = 4096 bytes
Sector size (logical/physical): 4096 bytes / 4096 bytes
I/O size (minimum/optimal): 32768 bytes / 1048576 bytes
Disklabel type: gpt
Disk identifier: 20175BB3-3579-1049-880E-F1012005D754

Device       Start     End Sectors  Size Type
/dev/sdb1     2048 1292287 1290240  4.9G Solaris /usr & Apple ZFS
/dev/sdb9  1292288 1308671   16384   64M Solaris reserved 1
```
Until error `GPT PMBR size mismatch` exists partprobe is failing to reload the partitions from disk to kernel.

Reply:
do you know what is the special thing that "zpool online -e <pool_name> <disk_name>" does which is making this error go away?

Reply:
I think relabeling should be helped here...
https://github.com/openebs/cstor/blob/4a47a5fe7b29f1d81bce3471be2982f0a12d24d8/lib/libzfs/libzfs_pool.c#L2411

Reply:
@Sai Chaithanya after step 1, the zfs will resize the pool. you might need to do partprobe to make kernel aware of it. You should not need 3rd step again to do resize.